src/group/libp2p_group_gossip_server.erl

```diff
@@ -1,105 +1,214 @@
 
-handle_call(workers, _From, State=#state{}) ->
-    {reply, connections(all, State), State};
 handle_call(Msg, _From, State) ->
     lager:warning("Unhandled call: ~p", [Msg]),
     {reply, ok, State}.
 
+handle_cast({add_handler, Key, Handler}, State=#state{handlers=Handlers}) ->
+    {noreply, State#state{handlers=maps:put(Key, Handler, Handlers)}};
+handle_cast({remove_handler, Key}, State=#state{handlers=Handlers}) ->
+    {noreply, State#state{handlers=maps:remove(Key, Handlers)}};
+
+handle_cast({request_target, inbound, WorkerPid}, State=#state{}) ->
+    {noreply, stop_inbound_worker(WorkerPid, State)};
 handle_cast({request_target, Kind=peerbook, WorkerPid}, State=#state{tid=TID}) ->
-    PeerAddrs = [ libp2p_crypto:address_to_p2p(Key)
-                  || Key <- libp2p_peerbook:keys(libp2p_swarm:peerbook(TID)) ],
+    PeerKeys = try
+                   libp2p_peerbook:keys(libp2p_swarm:peerbook(TID))
+               catch
+                   _:_ -> []
+               end,
+    PeerAddrs = [ libp2p_crypto:address_to_p2p(Key) || Key <- PeerKeys ],
     %% Get currently connected addresses
-    {CurrentAddrs, _} = lists:unzip(connections(Kind, State)),
+    {CurrentAddrs, _} = lists:unzip(connections(all, State)),
     LocalAddr = libp2p_crypto:address_to_p2p(libp2p_swarm:address(TID)),
     %% Exclude the local swarm address from the available addresses
     ExcludedAddrs = CurrentAddrs ++ [LocalAddr],
     %% Remove the current addrs from all possible peer addresses
     TargetAddrs = sets:to_list(sets:subtract(sets:from_list(PeerAddrs),
                                              sets:from_list(ExcludedAddrs))),
     {noreply, assign_target(Kind, WorkerPid, TargetAddrs, State)};
-handle_cast({request_target, Kind=seed, WorkerPid}, State=#state{seed_nodes=SeedAddrs}) ->
-    {CurrentAddrs, _} = lists:unzip(connections(Kind, State)),
-    TargetAddrs = sets:to_list(sets:subtract(sets:from_list(SeedAddrs), sets:from_list(CurrentAddrs))),
+handle_cast({request_target, Kind=seed, WorkerPid}, State=#state{tid=TID, seed_nodes=SeedAddrs}) ->
+    {CurrentAddrs, _} = lists:unzip(connections(all, State)),
+    LocalAddr = libp2p_crypto:address_to_p2p(libp2p_swarm:address(TID)),
+    %% Exclude the local swarm address from the available addresses
+    ExcludedAddrs = CurrentAddrs ++ [LocalAddr],
+    TargetAddrs = sets:to_list(sets:subtract(sets:from_list(SeedAddrs),
+                                             sets:from_list(ExcludedAddrs))),
     {noreply, assign_target(Kind, WorkerPid, TargetAddrs, State)};
-handle_cast({send, Bin}, State=#state{}) ->
+handle_cast({send, Key, Data}, State=#state{}) ->
     {_, Pids} = lists:unzip(connections(all, State)),
-    lists:foreach(fun(Pid) ->
-                          libp2p_group_worker:send(Pid, ignore, Bin)
-                  end, Pids),
-    {noreply, State};
-handle_cast({send_ready, _Ref, _Ready}, State=#state{}) ->
+    case (catch libp2p_gossip_stream:encode(Key, Data)) of
```

Comment:
If Key and Data are binaries, how can this fail?

Reply:
Key is a string and Data is a binary. I guess it could fail if you passed in ones that aren't?

Reply:
Added a comment to that effect (safe guarding the gossip server from bad input)